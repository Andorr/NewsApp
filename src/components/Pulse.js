// @flow
import * as React from 'react';
import className from 'classnames';
import {withStyles} from '@material-ui/core/styles';

// Project Components
import Flex from './layout/Flex';

type P = {
    classes: Object,
    children: ?any,
    className: string,
    label: ?string,
}

const styles: Object = {
    root: {
        backgroundColor: 'rgb(151, 41, 41)',
    },
    square: {
        borderRadius: 0,
    },
    content: {
        padding: 8,
        
        color: 'white',
    }
}

const Pulse: React.StatelessFunctionalComponent<P> = (props) => {
    const {classes} = props;
    return (
        <div className={className(classes.root, 'pulse', (props.label || props.children) ? classes.square : '', props.className)}>
            {(props.label || props.children) &&
                <Flex className={classes.content}>
                    {props.label}
                    {props.children}
                </Flex>
            }
        </div>
    )
}

export default withStyles(styles)(Pulse);