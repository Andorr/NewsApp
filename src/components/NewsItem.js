// @flow
import React, {Component} from 'react';
import {withStyles} from '@material-ui/core/styles';
import classNames from 'classnames';
import {Link} from 'react-router-dom';

// Material UI Components
import Typography from '@material-ui/core/Typography';
import Paper from '@material-ui/core/Paper';

// Project Components
import Pulse from '../components/Pulse';
import Flex from '../components/layout/Flex';

type P = {
    classes: Object,
    className: string,
    image?: string,
    title?: string,
    subtitle?: string,
    large: boolean,
    to: string,

    highlight: ?string,
}

const styles: Object = {
    root: {
        overflow: 'hidden',
        backgroundColor: 'white',
        width: '100%',
    },
    image: {
        width: '100%',
        height: 'auto',
        maxHeight: 275,
        objectFit: 'cover',
    },
    large: {
        maxHeight: 400,
    },
    imageWrapper: {
        backgroundColor: 'whitesmoke',
        position: 'relative',
        minHeight: 100,
        '@media only screen and (max-width: 600px)': {
            minHeight: 100,
        }
    },
    textWrapper: {
        padding: 28,
        flexGrow: 1,
    },
    text: {
        fontSize: '1.8rem',
    },
    lgText: {
        fontSize: '3rem',
    },
    pulse: {
        position: 'absolute',
        bottom: '-14px',
        height: 'auto',
        right: 30,
    }
}

class NewsItem extends Component<P> {

    render() {
        const {classes, className, image, title, subtitle} = this.props;
        return (
            <Paper className={classNames(classes.root, className)} square elevation={1}>
                <Link to={this.props.to} style={{textDecoration: 'none'}}>
                    <div className={classes.imageWrapper}>
                        <img className={classNames(classes.image, this.props.large ? classes.large : '')} src={image} alt={title} />
                        {this.props.highlight && <Pulse className={classes.pulse} label={this.props.highlight}/>}
                    </div>
                    <div className={classes.textWrapper}>
                        <Typography className={classNames(classes.text, this.props.large ? classes.lgText : '')} gutterBottom >{title}</Typography>
                        {this.props.large && <Typography variant='title' >{subtitle}</Typography>}
                    </div>
                </Link>
            </Paper>
        )
    }
}

 NewsItem.defaultProps = {
    to: '/',
 }

export default withStyles(styles)(NewsItem);