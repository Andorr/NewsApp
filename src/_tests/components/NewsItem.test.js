import React from 'react';
/* import {shallow, mount} from 'enzyme'; */
import { createShallow , createMount, type ReactWrapper } from '@material-ui/core/test-utils';

// Components
import NewsItem from '../../components/NewsItem';
import Pulse from '../../components/Pulse';
import Typography from '@material-ui/core/Typography';

describe('NewsItem component', () => {
    let shallow: ShallowWrapper;
    let mount: ShallowWrapper;
    
    beforeEach(() => {
        shallow = createShallow(); // Necassary for Material-UI's "withStyles()"
        mount = createMount();
    });

    afterEach(() => {
        mount.cleanUp();
    });

    it('[NewsItem] Testing render', () => {
        const component: ShallowWrapper = shallow(<NewsItem />);
        expect(component).toMatchSnapshot();
    });

    // Testing title props
    it('[NewsItem] Testing props: "Title"', () => {
        const title: string = 'Hello Test';

        const component: ShallowWrapper = shallow(<NewsItem title={title} />).dive();
        expect(component.find(Typography).findWhere(x => x.text() === title).length).toBe(1);
    });

    // Testing subtitle props
    it('[NewsItem] Testing props: "Subtitle"', () => {
        const subtitle: string = 'Hello Test';

        const component: ShallowWrapper = shallow(<NewsItem subtitle={subtitle} />).dive();
        expect(component.find(Typography).findWhere(x => x.text() === subtitle).length).toBe(1);
    });

    // Testing image props
    it('[NewsItem] Testing props: "Image"', () => {
        const IMAGE_PATH: string = 'https://MYIMAGE.com/';

        const component: ShallowWrapper = shallow(<NewsItem image={IMAGE_PATH} />).dive();
        expect(component.find('img').first()).toBeDefined();
        expect(component.find('img').first().prop('src')).toEqual(IMAGE_PATH);
    });

    // Testing highlight props
    it('[NewsItem] Testing props: "Highlight"', () => {
        // Should not render a Pulse component
        const noHighlightComp: ShallowWrapper = shallow(<NewsItem />).dive();
        expect(noHighlightComp.find(Pulse).length).toBe(0);

        // Should render a Pulse component
        const highlightText: string = 'Hello';
        const highlightComp: ShallowWrapper = shallow(<NewsItem highlight={highlightText}/>).dive();
        expect(highlightComp.find(Pulse).length).toBe(1);

        // Check if Pulse component has correct props label
        expect(highlightComp.find(Pulse).prop('label')).toEqual(highlightText);
    });
});