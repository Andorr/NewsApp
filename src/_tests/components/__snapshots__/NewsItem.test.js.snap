// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsItem component [NewsItem] Testing render 1`] = `
<NewsItem
  classes={
    Object {
      "image": "NewsItem-image-2",
      "imageWrapper": "NewsItem-imageWrapper-4",
      "large": "NewsItem-large-3",
      "lgText": "NewsItem-lgText-7",
      "pulse": "NewsItem-pulse-8",
      "root": "NewsItem-root-1",
      "text": "NewsItem-text-6",
      "textWrapper": "NewsItem-textWrapper-5",
    }
  }
/>
`;
